### ISSUE TITLE
Manual Target Submission API Endpoint & Control Panel Status Retrieval

### ISSUE THREAD

**Issue by jvillemare** [FlintHill/SUAS-Competition#68]
**Summary**

Add more API endpoints to the flask app to: Receive a manual crop target submission _and_ allow for the control panel items to have retrievable statuses.

**Purpose**

To finish and complete the Manual Target Submitter portion of the GCI Web UI _and_ make the control panel display current and accurate information on the GCS subprocesses.

### **Specifics for the Manual Target Submission API Endpoint** ###

What the flask app needs to be able to also receive—and what I neglected to add in Issue #67—is to be able to receive the MTS crop data.

[The code that submits the crop data can be seen here.](https://github.com/FlintHill/SUAS-Competition/blob/gcs-web-ui-start/scripts/gcs/static/rsc/js/main.js#L293-L335)

Here is the basic specs of the new MTS endpoint that needs to be added:

_URL:_ `/post/target`

_Data Request Type:_ JSON

_Data Request Example:_ 
```
{ 
	targetTopLeftX: 30, 
	targetTopLeftY: 20, 

	targetBottomRightX: 50,
	targetBottomRightY: 40,

	imageFilename: "0400292.jpg"
}
```

_Raw Data Request Example:_

![screen shot 2017-10-17 at 9 49 03 am](https://user-images.githubusercontent.com/6162516/31668743-0ab890a8-b321-11e7-84cc-52875ee884d2.png)

_Data Request Specification:_

- **integer targetTopLeftX (pixel):** 
   - The top left X point of the crop in imageFilename image. 
- **integer targetTopLeftY (pixel):**
   - The top left Y point of the crop in imageFilename image. 
- **integer targetBottomRightX (pixel):**
   - The bottom right X point of the crop in imageFilename image.
- **integer targetBottomRightY (pixel):**
   - The bottom right Y point of the crop in imageFilename image.
- **String imageFilename:**
   - The filename of the image that is the target of the crop.

_Data Request Assumptions:_

- imageFilename will always include the extension.
- Because the aspect ratio of the crop`targetBottomRightX - targetTopLeftX = targetBottomRightY - targetTopLeftY = extrude`

_Data Response:_

The data returned by the server can be like the responses described in #67 which is either `{status: "success"}` or `{status: "failure"}`. 

---

### **Specifics for Control Panel Status Retrieval** ###

[As I've posted in my comment in Issue 67](https://github.com/FlintHill/SUAS-Competition/issues/67#issuecomment-336924121), this is what the Control Panel screen looks like:

![copied image](https://user-images.githubusercontent.com/6162516/31620394-26f29438-b265-11e7-89e2-612e3158169e.png)

The power button, or turn on/off, toggable button, controls are sent to the API endpoints `/post/sda/<string>` and `/post/img_proc/<string>`, for example. To keep conventions the same and to limit the number of distinct API endpoint URLs, we should simply add a variation to existing URLs to retrieve the subprocess status, such as:

- `/get/sda`
- `/get/img_proc`
- `/get/interop`

All endpoints should respond with a basic JSON array, that looks like either:

_In the case of `/get/img_proc` or `/get/sda`:_

```
{
  "status": "connected",
  "runtime": "4 mins, 3 secs"
}
```

or, _in the case of `get/interop`:_

```
{
  "status": "connected",
  "emergent_position": "[38.783, -77.312]",
  "airdrop_position": "[38.783, -77.312]",
  "off-axis_position": "[38.783, -77.312]"
}
```

---

**Comment by DarkmatterVale** [comment]
The endpoint ```/post/target``` needs to also send the target crop information that the user selects.

**Comment by jvillemare** [comment]
@DarkmatterVale, at the top of this issue, the Data Request Type, Data Request Example, Raw Data Request Example, Data Request Specification, and Data Request Assumptions subsections detail the _request_ that will be made by my script from the web browser, _not_ the response that is expected.

This is the example given of what will be sent to the Flask backend:

```
{ 
	targetTopLeftX: 30, 
	targetTopLeftY: 20, 

	targetBottomRightX: 50,
	targetBottomRightY: 40,

	imageFilename: "0400292.jpg"
}
```

This is an example of what is expected to be returned from the Flask app:

`{status: "success"}`, or `{status: "failure"}`.

**Comment by jvillemare** [comment]
I just talked with you, @DarkmatterVale, and I now realize that you mean the orientation, shape, shape color, letter, and letter color information of the actual cropped target. My mistake, I'll commit it, ASAP.

**Comment by jvillemare** [comment]
@DarkmatterVale , the target data submission problem has now been fixed in commit https://github.com/FlintHill/SUAS-Competition/commit/8c995a4878002bb1ddc7d719097c821affdbbec2 .

Here's a new a sample crop:

![screen shot 2017-10-17 at 3 04 56 pm](https://user-images.githubusercontent.com/6162516/31684259-18ece83c-b34d-11e7-9032-f70197bdf8da.png)

![screen shot 2017-10-17 at 3 05 20 pm](https://user-images.githubusercontent.com/6162516/31684262-1beda1de-b34d-11e7-8191-a87e1e1f4313.png)

_Here's the new crop data as it is submitted:_

![screen shot 2017-10-17 at 3 06 02 pm](https://user-images.githubusercontent.com/6162516/31684268-2436e42c-b34d-11e7-9029-ddbc1634fbf9.png)

**Comment by DarkmatterVale** [comment]
A few comments:

- ```/post/sda/<status>``` & ```/post/img_proc/<status>```: For these endpoints, you need to pass "disconnected" when they are disabled, and "connected" when they are enabled
- ```/get/interop```: I cannot pass the JSON values as a string, so the result will look more or less like the following:
```
{
                "status": "disconnected",
                "emergent_position": [0, 0],
                "airdrop_position": [0, 0],
                "off-axis_position": [0, 0]
}
```

Is that alright?
- ```/get/imgs```: Should this be replaced with ```/get/available_images```?
- ```/imgs/<image>```: Should this be replaced with ```/get/imgs/<image>```?

Also, I get a messed up UI when I go to the page, like below:

![screen shot 2017-10-17 at 8 38 40 pm](https://user-images.githubusercontent.com/6094957/31695909-4d679052-b37b-11e7-8f8c-1c680d5356fc.png)

I believe this stems from a CSS file not being found?

![screen shot 2017-10-17 at 8 39 59 pm](https://user-images.githubusercontent.com/6094957/31695918-5e7961f4-b37b-11e7-8d2a-a7453ae162b6.png)

**Comment by jvillemare** [comment]
**In response to your comments in the order that they appear in:**

### `/post/sda/<status> & /post/img_proc/<status> endpoints...` ###

After talking to you, I fixed this in commit https://github.com/FlintHill/SUAS-Competition/commit/f23085961a4d768fda4f29db1771c75926f65244 .

### `/get/interop: I cannot pass the JSON values as a string...` ###

This is fine. I can simply use a JQuery version of toString() to print the two-index array as-is.

### `Is that alright?.../get/imgs...` ###

With the images API endpoint, to be keep conventions consistent and simple, we should just have:

- `/get/imgs` to return a list of available images, and
- `/get/imgs/<image filename>` to return a specific image.

### `I get a messed up UI...stems from a CSS file not being found?` ###

Not entirely. You're able to see all the Materialize CSS buttons, navbar, dropdowns, etc.. So, if your error message is correct, then the only resource missing is the imageareaselect library.

I fixed this by quickly adding the library—which I realized was not already included in the scripts/gcs/static/ directory because my working directory is seperate from where the repository is on my computer—by _committing a quick patch, see:_ https://github.com/FlintHill/SUAS-Competition/commit/0ab6d044be25bdae0663f47411f31536cfd4a59a .

**Comment by DarkmatterVale** [comment]
More changes to report:

1) APIs have been changed. You should be able to reach all of the APIs now. The interop post  target API needs more work, but the behavior you will see in the web client is exactly as it will be in production
2) I'm still getting the weird formatting error among others. This is only a problem in Safari though, so we just need to use chrome at competition (don't worry about fixing safari).
3) I cannot turn off any of the components --> Once I start SDA or Image Processing, they always keep running
4) When connected to SDA or image processing, the UI doesn't change the big red "Disconnected" box to a green "Connected" box
4) I can still toggle the interop connection
5) An error occurs every time I click "Load Images":

```jquery-3.2.1.min.js:4 GET http://localhost:5000/imgs/get.php 404 (NOT FOUND)```

**Comment by DarkmatterVale** [comment]
But all of the APIs have been added so you can probably close this issue.

**Comment by DarkmatterVale** [comment]
What we talked about that you need to fix:

1) Fix the toggle buttons
2) Change the tooltip for the power buttons to say "Toggle On/Off" instead of "Turn On"
3) Add zoom capability to image viewer (+ and - buttons are fine)
4) Fix error with runtime not being displayed
5) Fix error where "status" variable is displayed when I click the toggle button (when automatic refresh is not enabled)
6) Fix display issue within crop preview menu

I have made a change to the APIs. All requests except ```/get/imgs``` and ```/``` (index) return a ```request_status``` variable encoded in JSON. If the request was processed successfully and the code executed, "success" is returned. Otherwise, "failure" is returned.

**Comment by DarkmatterVale** [comment]
One additional change that needs to be made is that the coordinates for the airdrop location, emergent target, and off axis drop need to be displayed in a format other than one long string. I recommend you display them in a list format, like the following for each coordinate pair:

Lat: X
Lon: Y

**Comment by jvillemare** [comment]
**In response to** [your first comment since I've lasted posted in this issue:](https://github.com/FlintHill/SUAS-Competition/issues/68#issuecomment-337769956)

1. The APIs connect fine, and crop submission works on my end.
2. We can just use chrome at the competition—that's good. I haven't tested the web page for cross-browser compatibility, but generally, it should work on Firefox and/or Opera if it works on Chrome. The only minor concern is RAM usage, but Chrome is infamous for using RAM. But this problem is neither here nor there.
![screen shot 2017-10-20 at 10 45 21 pm](https://user-images.githubusercontent.com/6162516/31847348-74701510-b5e8-11e7-84bd-68bb6f10f7cb.png)
3. Processes being unable to turn/off was addressed and fixed in commit https://github.com/FlintHill/SUAS-Competition/commit/98b9a3d2b2dda7cf70eba461934c32f4242dc9b7 .
4. Status indicators not correctly being displayed was addressed and fixed in commits https://github.com/FlintHill/SUAS-Competition/commit/9fa3177586417b625a1a7c4f2a94240eb6f61585 and https://github.com/FlintHill/SUAS-Competition/commit/230d99415d31eeed74f5e5b5ba2772e8e8df07f2 . The statuses are now retrieved every time the power button is used.
5. Interop. connection power on/off button removed in commit https://github.com/FlintHill/SUAS-Competition/commit/98b9a3d2b2dda7cf70eba461934c32f4242dc9b7#diff-b361d09f31e3fbaeb7a32531087830e3 .
6. Images _were_ being loaded from a dummy script in a testing env. separate from my local SUAS-Competition repo. The URL that the images has now been corrected in line 151 of commit https://github.com/FlintHill/SUAS-Competition/commit/9fa3177586417b625a1a7c4f2a94240eb6f61585#diff-9ebef0815256d83bae1343312797d286 .

---

**In response to** [your second comment since I've lasted posted in this issue:](https://github.com/FlintHill/SUAS-Competition/issues/68#issuecomment-337770136)

I think we should still keep this issue open until we address the Manual Target Submission's zooming in and out feature that needs to be added, or, _should we open another issue for this?_

---

**In response to** [your third comment since I've lasted posted in this issue:](https://github.com/FlintHill/SUAS-Competition/issues/68#issuecomment-338017427)

1. Toggle buttons fixed in lines 400 to 420 in https://github.com/FlintHill/SUAS-Competition/commit/230d99415d31eeed74f5e5b5ba2772e8e8df07f2 .
2. Power button hover text fixed in https://github.com/FlintHill/SUAS-Competition/commit/3a7a77c35f93c1b98517655904477ab84c7e1802#diff-b361d09f31e3fbaeb7a32531087830e3R186 .
3. This is addressed in my response to your second response in the section above: _Should we open another issue for this?_
4. Process runtime displayed in https://github.com/FlintHill/SUAS-Competition/commit/230d99415d31eeed74f5e5b5ba2772e8e8df07f2#diff-9ebef0815256d83bae1343312797d286R395 .
5. Process status displayed in https://github.com/FlintHill/SUAS-Competition/commit/230d99415d31eeed74f5e5b5ba2772e8e8df07f2#diff-9ebef0815256d83bae1343312797d286R382 .
6. Formatting fixed with set table widths in https://github.com/FlintHill/SUAS-Competition/commit/3a7a77c35f93c1b98517655904477ab84c7e1802#diff-b361d09f31e3fbaeb7a32531087830e3R52 .

---

**In response to** [your fourth comment since I've lasted posted in this issue:](https://github.com/FlintHill/SUAS-Competition/issues/68#issuecomment-338069969)

I talked to you about this, and I've addressed this by putting the coordinates in the form: `[<lat>, <long>]` in commit https://github.com/FlintHill/SUAS-Competition/commit/230d99415d31eeed74f5e5b5ba2772e8e8df07f2#diff-9ebef0815256d83bae1343312797d286R388 .

---

**TL;DR:**

I fixed all the changes you requested, but, I still have one question, and that is: To address the Manual Target Submitter not supporting zoom-in and zoom-out—should that be a new issue, or included in this issue? I'd love to know your thoughts, @DarkmatterVale.

**Comment by DarkmatterVale** [comment]
Lets add a new issue, and I will close this one as soon as I verify all of your changes are good.

Thanks!

**Comment by DarkmatterVale** [comment]
I believe all issues have been resolved, so I will go ahead and close it. @GGJesus Please create a new issue whenever you get the chance for the zoom feature.
